### PR TITLE
Challenges

### PR DESCRIPTION
**Pull request by nickago**
This contains all core functionality and JSON API support, wanted to merge this in, so the community may judge harshly and swiftly. The next step for me is to get some the front end work done, and maybe even some testing *gasp*

### CODE REVIEW

**Comment by nickago** [comment]
Added from my dev branch: Frontend access to the challenges page, challenge prerequisites

**Comment by nickago** [comment]
Added from my dev branch: Documentation on the example

**Comment by nickago** [comment]
Going to add tests now, might have some redundancies with #9, I'll do an integrations check once one of the two are merged.

**Comment by nickago** [comment]
Added from my dev branch: Ability to serve files associated with the challenge

File: API.md [hunk 1]
```diff
@@ -12,6 +12,8 @@
   - [List teams you're invited to](#list-teams-youre-invited-to)
   - [Join a team](#join-a-team)
   - [Leave team](#leave-team)
+- [Challenges](#challenges)
```
Comment (nickfrostatx) [inline API.md]: nit: no leading space to hyphen

Comment (nickago) [inline API.md]: Issue seems to be resolved

**Pull request merged**